library/posture.c

```diff
@@ -0,0 +1,202 @@
+/*
+Copyright (c) 2019-2020 NetFoundry, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "posture.h"
+
+#include <utils.h>
+
+#define NANOS(s) ((s) * 1e9)
+#define MILLIS(s) ((s) * 1000)
+
+static void ticker_cb(uv_timer_t *t);
+
+static void free_string_arr(char **string_arr, int size);
+
+static void ziti_handle_mac(ziti_context ztx, char *id, char **mac_addresses, int num_mac);
+
+static void ziti_handle_domain(ziti_context ztx, char *id, char *domain);
+
+static void ziti_handle_os(ziti_context ztx, char *id, char *os_type, char *os_version, char *os_build);
+
+static void
+ziti_handle_process(ziti_context ztx, char *id, char *path, bool is_running, char *sha_512_hash, char **signers,
+                    int num_signers);
+
+extern void posture_init(struct ziti_ctx *ztx, long interval_secs) {
+    if (ztx->posture_checks == NULL) {
+        NEWP(pc, struct posture_checks);
+        pc->interval = (double) interval_secs;
+        ztx->posture_checks = pc;
+    }
+
+    if (!uv_is_active((uv_handle_t *) &ztx->posture_checks->timer)) {
+        uv_timer_init(ztx->loop, &ztx->posture_checks->timer);
+        ztx->posture_checks->timer.data = ztx;
+        uv_timer_start(&ztx->posture_checks->timer, ticker_cb, MILLIS(interval_secs), MILLIS(interval_secs));
+        uv_unref((uv_handle_t *) &ztx->posture_checks->timer);
+    }
+}
+
+extern void ziti_posture_checks_free(struct posture_checks *pcs) {
+    FREE(pcs);
+}
+
+static void ticker_cb(uv_timer_t *t) {
+    struct ziti_ctx *ztx = t->data;
+    ziti_send_posture_data(ztx);
+}
+
+struct query_info {
+    ziti_service *service;
+    ziti_posture_query_set *query_set;
+    ziti_posture_query *query;
+};
+
+void ziti_send_posture_data(struct ziti_ctx *ztx) {
+    ZITI_LOG(DEBUG, "starting to send posture data");
+    const char *service_id;
+
+    NEWP(domainInfo, struct query_info);
+    NEWP(osInfo, struct query_info);
+    NEWP(macInfo, struct query_info);
+    struct query_info *procInfo = NULL;
+
+    NEWP(processes, model_map);
```

Comment:
any reason to allocate vs using a stack variable?

Reply:
Fixed.